<a href="https://colab.research.google.com/github/MattWroclaw/neural-networks/blob/main/03_keras/01_kasyfikacja_binarna_wieloklasowa_regresja.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Model sekwencjny (Sequential Model):
1. [Wprowadzenie - pierwszy model](#a0)
2. [Funkcje aktywacji](#a1)
3. [Kompilacja modelu](#a2)
4. [Trenowanie modelu](#a3)
5. [Przykład - klasyfikacja binarna](#a4)
6. [Przykład - klasyfikacja wieloklasowa](#a5)
7. [Przykład - regresja](#a6)



Konfiguracja środowiska


In [ ]:
# Przygotowanie środowiska do pracy z Tensorflow 2.0.
# Jeśli otrzymasz błąd podczas instalacji Tensorflow uruchom tę komórkę raz jeszcze.

!pip uninstall -y tensorflow
!pip install -q tensorflow==2.0.0

In [1]:
import tensorflow as tf
import numpy as np
import pandas as pd
import plotly.express as px
tf.__version__

'2.17.0'

### <a name='a0'></a>Wprowadzenie - pierwszy model
Model sekwencyjny to nic innego jak liniowy stos warstw.

In [2]:
# utworzenie instancji klasy Sequential
from tensorflow.keras.models import Sequential

model = Sequential()
print(model)

<Sequential name=sequential, built=False>


Podstawowym elementem składowym modelu są warstwy. Aby dodać najbardziej standardową warstwę - warstwę gęsto połączoną należy użyć warstwy **Dense**. Aby dodać warstwę do modelu należy użyć metody *.add()*


In [3]:
from tensorflow.keras.layers import Dense

model.add(Dense(units=4, input_shape=(10,)))

/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Wyświetlenie podsumowania modelu: metoda *.summary()*

In [4]:
model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ dense (Dense)                        │ (None, 4)                   │              44 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 44 (176.00 B)

 Trainable params: 44 (176.00 B)

 Non-trainable params: 0 (0.00 B)

Dodanie kolejnej warstwy

In [5]:
model.add(Dense(units=2))

model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ dense (Dense)                        │ (None, 4)                   │              44 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_1 (Dense)                      │ (None, 2)                   │              10 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 54 (216.00 B)

 Trainable params: 54 (216.00 B)

 Non-trainable params: 0 (0.00 B)

### <a name='a1'></a>Funkcje aktywacji
Istotnym elemenem sieci neuronowych jest dobór odpowiednich funkcji aktywacji. Funkcje aktywacji jak sama nazwa wskazuje są odpowiedzialne za aktywowanie odpowiednich neuronów podczas procesu uczenia.

Jeżeli nie określimy podczas dodawania warstwy funkcji aktywacji, domyślnie stosowana jest liniowa funkcja aktywacji, tzn. $a(x)=x$

Warstwa z liniową funkcją aktywacji może uczyć się tylko liniowych przekształceń danych wejściowych. Dlatego stosuje się różne funkcje aktywacji aby rozwiazywać problemy nieliniowe.

[Keras: Funkcje aktywacji](https://keras.io/activations/)

In [6]:
from tensorflow.keras.activations import linear

random_data = sorted(np.random.randn(200))
data = pd.DataFrame({'data': random_data, 'linear': linear(random_data)})
data.head()

data    linear
0 -2.639736 -2.639736
1 -2.358621 -2.358621
2 -2.131749 -2.131749
3 -2.105261 -2.105261
4 -2.064501 -2.064501

In [7]:
px.line(data, x='data', y='linear', width=800, range_y=[-2, 2])

In [8]:
from tensorflow.keras.activations import sigmoid

data = pd.DataFrame({'data': random_data, 'sigmoid': sigmoid(random_data)})
data.head()

data   sigmoid
0 -2.639736  0.066624
1 -2.358621  0.086383
2 -2.131749  0.106049
3 -2.105261  0.108586
4 -2.064501  0.112595

In [9]:
px.line(data, x='data', y='sigmoid', width=800, range_y=[-0.5, 1.5])

In [10]:
from tensorflow.keras.activations import relu

data = pd.DataFrame({'data': random_data, 'relu': relu(random_data)})
data.head()

data  relu
0 -2.639736   0.0
1 -2.358621   0.0
2 -2.131749   0.0
3 -2.105261   0.0
4 -2.064501   0.0

In [11]:
px.line(data, x='data', y='relu', width=800, range_y=[-0.5, 1.5])

In [12]:
from tensorflow.keras.activations import tanh

data = pd.DataFrame({'data': random_data, 'tanh': tanh(random_data)})
data.head()

data      tanh
0 -2.639736 -0.989861
1 -2.358621 -0.982279
2 -2.131749 -0.972245
3 -2.105261 -0.970757
4 -2.064501 -0.968312

In [13]:
px.line(data, x='data', y='tanh', width=800, range_y=[-1.5, 1.5])

In [14]:
model = Sequential()
model.add(Dense(units=8, activation='relu', input_shape=(10,))) # warstwa gęsta z 8 neuronami, 10 - ilość klas
model.add(Dense(units=1, activation='sigmoid'))
model.summary()

/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:87: UserWarning:

Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.



Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ dense_2 (Dense)                      │ (None, 8)                   │              88 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_3 (Dense)                      │ (None, 1)                   │               9 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 97 (388.00 B)

 Trainable params: 97 (388.00 B)

 Non-trainable params: 0 (0.00 B)

### <a name='a2'></a>Kompilacja modelu
Przed rozpoczęciem trenowania sieci należy odpowiednio skonfigurować proces uczenia. W tym kroku określamy:
* rodzaj optymalizatora ([Keras - Optymalizatory](https://keras.io/optimizers/))
* funkcję straty ([Keras - Funkcje Straty](https://keras.io/losses/))
* metryki, które będziemy obserwować podczas trenowania sieci ([Keras - Metryki](https://keras.io/metrics/))

In [15]:
# klasyfikacja binarna
model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

# klasyfikacja wieloklasowa
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# regresja
model.compile(optimizer='rmsprop',
              loss='mse')

### <a name='a3'></a>Trenowanie modelu
Za dane wejściowe do modelu należy przekazać Numpy arrays:
* **epochs** - krotność przejścia danych przez sieć w procesie uczenia
* **batch_size** - rozmiar wsadu po którym następuje aktualizacja wag
* **validation_split** - część danych treningowych, które zostaną wykorzystane jako zbiór walidacyjny
* **validation_data** - (x_val, y_val) - dane wykorzystane do walidacji modelu

In [ ]:
# model.fit(data, labels, epochs=10, batch_size=32)
# model.fit(data, labels, epochs=10, batch_size=32, validation_split=0.2)
# model.fit(data, labels, epochs=10, batch_size=32, validation_data=(x_val, y_val))

### <a name='a4'></a> Przykład - klasyfikacja binarna

In [16]:
data = np.random.randn(1000, 150)
labels = np.random.randint(2, size=(1000, 1))

print(data.shape)
print(labels.shape)

(1000, 150)
(1000, 1)


Wyjaśnienie krok po kroku:

np.random.randint(2, size=(1000, 1)): Ta część kodu używa funkcji np.random.randint z biblioteki NumPy do generowania liczb losowych.
2: Argument ten określa, że liczby losowe będą generowane z zakresu od 0 do 1 (włącznie). W klasyfikacji binarnej zazwyczaj używa się 0 i 1 do reprezentowania dwóch klas.
size=(1000, 1): Ten argument określa rozmiar tablicy, która ma zostać utworzona. W tym przypadku zostanie utworzona tablica o wymiarach 1000 wierszy i 1 kolumnie. Oznacza to, że zostanie wygenerowanych 1000 etykiet.
labels = ...: Wygenerowana tablica liczb losowych jest przypisywana do zmiennej labels.

In [17]:
labels[:10]

array([[0],
       [0],
       [1],
       [1],
       [1],
       [1],
       [0],
       [0],
       [0],
       [1]])

In [18]:
model = Sequential()
model.add(Dense(units=32, activation='relu', input_shape=(150,)))
model.add(Dense(1, activation='sigmoid'))

model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

model.fit(data, labels, epochs=20)

Epoch 1/20
32/32 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.4960 - loss: 0.7900   
Epoch 2/20
32/32 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.5632 - loss: 0.6945
Epoch 3/20
32/32 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.6464 - loss: 0.6313 
Epoch 4/20
32/32 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.6753 - loss: 0.6071 
Epoch 5/20
32/32 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.7160 - loss: 0.5720
Epoch 6/20
32/32 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.7710 - loss: 0.5349
Epoch 7/20
32/32 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.7817 - loss: 0.5118 
Epoch 8/20
32/32 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.7843 - loss: 0.4936 
Epoch 9/20
32/32 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.8198 - loss: 0.4657 
Epoch 10/20
32/32 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.8390 - loss: 0.4415 
Epoch 11/20
32/32 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.8565 - loss: 0.4223 
Epoch 12/20
32/32 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0

In [19]:
model = Sequential()
model.add(Dense(units=32, activation='relu', input_shape=(150,)))
model.add(Dense(1, activation='sigmoid'))

model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

model.fit(data, labels, epochs=20, batch_size=30) # po 30stej próbce będzie walidacja wagi

Epoch 1/20
34/34 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.4766 - loss: 0.8543   
Epoch 2/20
34/34 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.5623 - loss: 0.7172 
Epoch 3/20
34/34 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.6180 - loss: 0.6534 
Epoch 4/20
34/34 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.6932 - loss: 0.5857 
Epoch 5/20
34/34 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.7132 - loss: 0.5670 
Epoch 6/20
34/34 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.7373 - loss: 0.5348 
Epoch 7/20
34/34 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7784 - loss: 0.5074
Epoch 8/20
34/34 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.7940 - loss: 0.4871 
Epoch 9/20
34/34 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.8336 - loss: 0.4503 
Epoch 10/20
34/34 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.8584 - loss: 0.4161 
Epoch 11/20
34/34 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.8717 - loss: 0.4020 
Epoch 12/20
34/34 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy:

In [20]:
model = Sequential()
model.add(Dense(units=32, activation='relu', input_shape=(150,)))
model.add(Dense(1, activation='sigmoid'))

model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

model.fit(data, labels, epochs=20, batch_size=32, validation_split=0.2)
# 0,2 => dane do testów / 0,8 dane do nauki

Epoch 1/20
25/25 ━━━━━━━━━━━━━━━━━━━━ 1s 10ms/step - accuracy: 0.4674 - loss: 0.8464 - val_accuracy: 0.5700 - val_loss: 0.7547
Epoch 2/20
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5608 - loss: 0.7372 - val_accuracy: 0.5550 - val_loss: 0.7496
Epoch 3/20
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6185 - loss: 0.6537 - val_accuracy: 0.5400 - val_loss: 0.7493
Epoch 4/20
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6731 - loss: 0.6051 - val_accuracy: 0.5450 - val_loss: 0.7541
Epoch 5/20
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6887 - loss: 0.5669 - val_accuracy: 0.5200 - val_loss: 0.7613
Epoch 6/20
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7248 - loss: 0.5367 - val_accuracy: 0.5150 - val_loss: 0.7708
Epoch 7/20
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7434 - loss: 0.5162 - val_accuracy: 0.5350 - val_loss: 0.7820
Epoch 8/20
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.8255 - loss: 0.4558 - val_accuracy: 0.5300 - val_loss

In [21]:
model = Sequential()
model.add(Dense(units=32, activation='relu', input_shape=(150,)))
model.add(Dense(1, activation='sigmoid'))

model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

model.fit(data, labels, epochs=20, batch_size=32, validation_split=0.2, verbose=0)

In [22]:
model = Sequential()
model.add(Dense(units=32, activation='relu', input_shape=(150,)))
model.add(Dense(1, activation='sigmoid'))

model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

history = model.fit(data, labels, epochs=20, batch_size=32, validation_split=0.2, verbose=0)

Przypisanie do zmiennej *histiry* . Daje nam możliwość sprawdzenia dodatkowych danych

In [23]:
metrics = history.history
metrics.keys()

dict_keys(['accuracy', 'loss', 'val_accuracy', 'val_loss'])

In [24]:
model = Sequential()
model.add(Dense(units=32, activation='relu', input_shape=(150,)))
model.add(Dense(1, activation='sigmoid'))

model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

history = model.fit(data, labels, epochs=20, batch_size=32, validation_split=0.2, verbose=1)

Epoch 1/20
25/25 ━━━━━━━━━━━━━━━━━━━━ 1s 10ms/step - accuracy: 0.5076 - loss: 0.8274 - val_accuracy: 0.5100 - val_loss: 0.7527
Epoch 2/20
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5559 - loss: 0.7090 - val_accuracy: 0.5350 - val_loss: 0.7379
Epoch 3/20
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6364 - loss: 0.6637 - val_accuracy: 0.5350 - val_loss: 0.7330
Epoch 4/20
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6722 - loss: 0.6135 - val_accuracy: 0.5200 - val_loss: 0.7310
Epoch 5/20
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7165 - loss: 0.5692 - val_accuracy: 0.5500 - val_loss: 0.7311
Epoch 6/20
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7188 - loss: 0.5661 - val_accuracy: 0.5350 - val_loss: 0.7337
Epoch 7/20
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7900 - loss: 0.5191 - val_accuracy: 0.5500 - val_loss: 0.7394
Epoch 8/20
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.8195 - loss: 0.4911 - val_accuracy: 0.5450 - val_loss

GEnerujemy dane "rzeczywiste".

In [37]:
test_data = np.random.randn(15, 150)
test_labels = np.random.randint(2, size=(15, 1))

np.random.randint(2, ...): This part of the code uses the randint function from NumPy's random module to generate random integers.
The argument 2 specifies that the integers should be drawn from the range 0 (inclusive) to 2 (exclusive). This means it will only generate 0 or 1.
size=(5, 1): This argument specifies the shape of the output array. In this case, you're asking for an array with 5 rows and 1 column.



```
[[1]  
 [0]   
 [1]   
 [1]  
 [0]]
```



In [38]:
model.predict(test_data)

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 49ms/step


array([[0.12050517],
       [0.86438787],
       [0.6410461 ],
       [0.5906817 ],
       [0.3909368 ],
       [0.8175948 ],
       [0.9700837 ],
       [0.20972787],
       [0.28158206],
       [0.6008485 ],
       [0.58673954],
       [0.22685274],
       [0.7446513 ],
       [0.9364303 ],
       [0.06569249]], dtype=float32)

In [47]:
# model.predict_proba(test_data)

# Import necessary libraries
from tensorflow import keras

# Assuming your model is a binary classifier:
predictions = model.predict(test_data)
predicted_probabilities = predictions  # For binary classification, predictions are probabilities

# If you have a multi-class classifier, and you need probabilities for each class:
# and your model's output layer uses a softmax activation, you can access probabilities like this:
# predicted_probabilities = predictions

# If you need class labels, you can use argmax:
predicted_labels = predictions.argmax(axis=1)

print(predicted_probabilities)
print(predicted_labels)
# Note: This assumes your output layer uses 'sigmoid' for binary classification
# or 'softmax' for multi-class classification. If not, adjust accordingly.

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 34ms/step
[[0.12324271 0.15562774 0.12676243 0.07035228 0.07790477 0.1359667
  0.13492922 0.04915583 0.07710804 0.04895035]
 [0.14462627 0.15085153 0.12906404 0.14366671 0.14621621 0.03717632
  0.07472172 0.06107028 0.05793819 0.05466872]
 [0.09628847 0.18091062 0.24035503 0.07760637 0.09029748 0.04071181
  0.07869991 0.04014925 0.10893642 0.04604471]
 [0.08849732 0.09196696 0.23448393 0.144744   0.08910221 0.07422858
  0.11555141 0.04687843 0.08175761 0.0327895 ]
 [0.05665649 0.12874985 0.20217033 0.10953651 0.06219082 0.11987718
  0.127824   0.05097183 0.05275573 0.0892673 ]
 [0.11446884 0.08234062 0.05985812 0.08475245 0.04829307 0.1888871
  0.07890694 0.04234014 0.1428482  0.15730453]
 [0.07648019 0.08034172 0.07097768 0.07816881 0.23549172 0.1391375
  0.09228935 0.0733855  0.06865716 0.08507028]
 [0.09476407 0.31066033 0.08989897 0.05015648 0.14656578 0.02797841
  0.14070797 0.03626674 0.0369867  0.0660145 ]
 [0.12995183 0.14430872 0.09687762 0.07906362

Okay, let's dissect the line predicted_classes = np.argmax(predictions, axis=1) to understand what it does.

In essence:

This line of code finds the index of the maximum value along each row (axis 1) of the predictions array. This index represents the predicted class for each data sample.

Breakdown:

predictions: This is the output from your model's predict method. It's typically an array or a list of lists, where each inner list contains the probabilities of each class for a single data sample. For example, if your model has 3 classes, predictions might look like this for a single sample: [0.1, 0.7, 0.2].
np.argmax(...): This is a NumPy function that finds the index of the maximum value within an array. In our case, it's applied to the predictions array.
axis=1: This argument specifies the dimension along which argmax should operate. axis=1 means it will find the maximum value along each row of the predictions array. This is important because each row represents a single data sample, and we want to find the class with the highest probability for that sample.
How it determines the predicted class:

Let's say predictions for a single sample is [0.1, 0.7, 0.2]. When we apply np.argmax with axis=1, it will look for the maximum value within this row, which is 0.7. The index of this maximum value is 1. Therefore, np.argmax will return 1 as the predicted class for this sample.

Overall:

predicted_classes = np.argmax(predictions, axis=1) iterates through each row (data sample) in the predictions array, finds the index of the maximum value within that row (representing the class with the highest probability), and stores these indices in the predicted_classes array. This gives you the predicted class for each data sample.

Example:


predictions = [[0.1, 0.7, 0.2],
               [0.8, 0.1, 0.1],
               [0.3, 0.3, 0.4]]

predicted_classes = np.argmax(predictions, axis=1)
print(predicted_classes)  # Output: [1 0 2]
Use code with caution
In this example, the predicted classes are:

Sample 1: Class 1 (index 1)
Sample 2: Class 0 (index 0)
Sample 3: Class 2 (index 2)

### <a name='a5'></a> Przykład - klasyfikacja wieloklasowa

In [30]:
data = np.random.random((1000, 150))
labels = np.random.randint(10, size=(1000, 1))

In [31]:
print(data.shape)
print(labels.shape)

(1000, 150)
(1000, 1)


In [32]:
data[:3]

array([[0.86484328, 0.07596113, 0.77908867, 0.34740635, 0.45087249,
        0.6096593 , 0.04685915, 0.93623274, 0.18550294, 0.32111406,
        0.86654593, 0.22841423, 0.71287865, 0.80566297, 0.55716187,
        0.75353505, 0.62950397, 0.52715252, 0.42991572, 0.8444447 ,
        0.25716978, 0.36207481, 0.10754078, 0.85769845, 0.17391961,
        0.16175038, 0.17101295, 0.69463265, 0.11160273, 0.28191544,
        0.6067843 , 0.44495891, 0.79925579, 0.07645332, 0.46815551,
        0.33810784, 0.06186355, 0.27946658, 0.66965559, 0.14514803,
        0.78722615, 0.76170118, 0.64018944, 0.32343427, 0.32712218,
        0.27197648, 0.82266121, 0.44108471, 0.86979002, 0.0559398 ,
        0.93552878, 0.00528473, 0.8343252 , 0.45649992, 0.18790446,
        0.02011159, 0.83956446, 0.39844189, 0.42913447, 0.51913501,
        0.83443972, 0.76986163, 0.71091981, 0.02980957, 0.38125423,
        0.5606687 , 0.3989632 , 0.51339182, 0.1286984 , 0.51557876,
        0.89896249, 0.66195843, 0.34446515, 0.72

In [33]:
labels[:10]

array([[7],
       [0],
       [3],
       [7],
       [4],
       [4],
       [9],
       [7],
       [1],
       [7]])

In [34]:
from tensorflow.keras.utils import to_categorical
labels = to_categorical(labels, num_classes=10)
labels

array([[0., 0., 0., ..., 1., 0., 0.],
       [1., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 1.],
       [0., 0., 1., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 1., 0.]])

In [40]:
labels[1]

array([1., 0., 0., 0., 0., 0., 0., 0., 0., 0.])

In [41]:
model = Sequential()
model.add(Dense(units=32, activation='relu', input_shape=(150,)))
model.add(Dense(units=10, activation='softmax'))

model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

model.fit(data, labels, batch_size=32, epochs=30, validation_split=0.2)

Epoch 1/30


/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:87: UserWarning:

Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.



25/25 ━━━━━━━━━━━━━━━━━━━━ 1s 15ms/step - accuracy: 0.0971 - loss: 2.4414 - val_accuracy: 0.1350 - val_loss: 2.3354
Epoch 2/30
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.1000 - loss: 2.3327 - val_accuracy: 0.1050 - val_loss: 2.3212
Epoch 3/30
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.1083 - loss: 2.3130 - val_accuracy: 0.1200 - val_loss: 2.3288
Epoch 4/30
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.1299 - loss: 2.2879 - val_accuracy: 0.0900 - val_loss: 2.3275
Epoch 5/30
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.1582 - loss: 2.2757 - val_accuracy: 0.1150 - val_loss: 2.3380
Epoch 6/30
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.1338 - loss: 2.2633 - val_accuracy: 0.1050 - val_loss: 2.3393
Epoch 7/30
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.1325 - loss: 2.2567 - val_accuracy: 0.1200 - val_loss: 2.3311
Epoch 8/30
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.1806 - loss: 2.2410 - val_accuracy: 0.1050 - val_loss: 2.3372
Ep

In [42]:
test_data = np.random.random((10, 150))

model.predict(test_data)

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 49ms/step


array([[0.12324271, 0.15562774, 0.12676243, 0.07035228, 0.07790477,
        0.1359667 , 0.13492922, 0.04915583, 0.07710804, 0.04895035],
       [0.14462627, 0.15085153, 0.12906404, 0.14366671, 0.14621621,
        0.03717632, 0.07472172, 0.06107028, 0.05793819, 0.05466872],
       [0.09628847, 0.18091062, 0.24035503, 0.07760637, 0.09029748,
        0.04071181, 0.07869991, 0.04014925, 0.10893642, 0.04604471],
       [0.08849732, 0.09196696, 0.23448393, 0.144744  , 0.08910221,
        0.07422858, 0.11555141, 0.04687843, 0.08175761, 0.0327895 ],
       [0.05665649, 0.12874985, 0.20217033, 0.10953651, 0.06219082,
        0.11987718, 0.127824  , 0.05097183, 0.05275573, 0.0892673 ],
       [0.11446884, 0.08234062, 0.05985812, 0.08475245, 0.04829307,
        0.1888871 , 0.07890694, 0.04234014, 0.1428482 , 0.15730453],
       [0.07648019, 0.08034172, 0.07097768, 0.07816881, 0.23549172,
        0.1391375 , 0.09228935, 0.0733855 , 0.06865716, 0.08507028],
       [0.09476407, 0.31066033, 0.0898989

In [44]:
# model.predict_classes(test_data)

# Instead of predict_classes, use predict and argmax to get the class with the highest probability
predictions = model.predict(test_data)
predicted_classes = np.argmax(predictions, axis=1)
print(predicted_classes)

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 22ms/step
[1 1 2 2 2 5 4 1 4 1]


### <a name='a6'></a> Przykład - regresja

In [48]:
data = np.random.random((1000, 150))
labels = 50 * np.random.random(1000)

In [49]:
data[:3]

array([[6.77299202e-01, 4.55768032e-01, 5.96810266e-02, 9.15572584e-01,
        2.17799267e-01, 5.11021019e-01, 6.63482611e-02, 1.71991048e-01,
        7.02191463e-01, 9.10011085e-01, 9.02692499e-01, 9.38636896e-01,
        1.13006363e-02, 3.37562592e-01, 1.20427395e-01, 8.84805068e-02,
        6.65344923e-01, 5.21962995e-01, 4.15732778e-01, 1.38246273e-01,
        4.88457599e-01, 7.39181784e-02, 4.55456963e-02, 2.10112488e-01,
        8.00666235e-03, 5.29663077e-01, 6.21521405e-01, 8.12861911e-01,
        2.41554896e-01, 2.08855597e-01, 3.47343263e-01, 7.30406058e-01,
        4.24615684e-01, 5.87134596e-01, 9.50080522e-02, 6.64235450e-01,
        8.41114063e-01, 1.23651947e-03, 7.92614260e-01, 9.70866715e-01,
        9.03196379e-01, 2.89828031e-01, 3.71973250e-01, 3.24278610e-01,
        8.42038461e-01, 3.30147989e-01, 4.08877538e-01, 6.02613225e-01,
        7.83026921e-01, 2.87304113e-01, 1.71522401e-01, 8.85191915e-01,
        5.96672901e-01, 7.01136732e-01, 4.28796298e-01, 2.562664

In [50]:
labels[:10]

array([40.83488963, 13.54614499,  9.82216082, 39.98487433, 35.23886242,
       24.98660607, 24.12461553, 11.01770545, 42.80980729, 36.91008471])

In [51]:
model = Sequential()
model.add(Dense(units=32, activation='relu', input_shape=(150,)))
model.add(Dense(units=1))

model.compile(optimizer='rmsprop',
              loss='mse')

model.fit(data, labels, epochs=30, batch_size=32, validation_split=0.2)

Epoch 1/30


/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:87: UserWarning:

Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.



25/25 ━━━━━━━━━━━━━━━━━━━━ 1s 8ms/step - loss: 627.2679 - val_loss: 415.8598
Epoch 2/30
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 355.8130 - val_loss: 245.6504
Epoch 3/30
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 222.9471 - val_loss: 189.2969
Epoch 4/30
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 196.9194 - val_loss: 185.9766
Epoch 5/30
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 186.8803 - val_loss: 185.2614
Epoch 6/30
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 183.8979 - val_loss: 185.4181
Epoch 7/30
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 193.7612 - val_loss: 185.6388
Epoch 8/30
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 195.6472 - val_loss: 186.4591
Epoch 9/30
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 196.5423 - val_loss: 185.8080
Epoch 10/30
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 182.4408 - val_loss: 185.7743
Epoch 11/30
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 196.1996 - val_loss: 185.9697
Epoch 12/30
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s

Dlaczego w modelu regresji nie podajemy "metrics" do metody model.compile ?

Gemini
W modelu regresji nie podajemy "metrics" takich jak "accuracy" do metody model.compile, ponieważ "accuracy" nie jest odpowiednią miarą do oceny wydajności modeli regresji.

Wyjaśnienie:

Accuracy jest miarą używaną w klasyfikacji, gdzie liczy się odsetek poprawnie sklasyfikowanych próbek. W regresji natomiast przewidujemy ciągłe wartości, a nie kategorie.
W regresji chcemy ocenić, jak blisko przewidywania są rzeczywistych wartości. Do tego celu używamy innych metryk, takich jak:
Mean Squared Error (MSE): Średni kwadrat błędu między przewidywaniami a wartościami rzeczywistymi.
Root Mean Squared Error (RMSE): Pierwiastek kwadratowy z MSE.
Mean Absolute Error (MAE): Średnia wartość bezwzględna błędu między przewidywaniami a wartościami rzeczywistymi.
R-squared: Współczynnik determinacji, który mierzy, jak dobrze model pasuje do danych.
Przykład:


from tensorflow import keras

# ... definicja modelu ...

model.compile(optimizer='adam',
              loss='mse',  # Funkcja straty dla regresji
              metrics=['mae', 'rmse']) # Metryki dla regresji
Use code with caution
W powyższym przykładzie używamy mse jako funkcji straty (loss function) i mae oraz rmse jako metryk do oceny modelu regresji.

Podsumowanie:

"Accuracy" nie jest odpowiednią miarą dla modeli regresji.
W regresji używamy metryk takich jak MSE, RMSE, MAE i R-squared do oceny wydajności modelu.
Metryki te mierzą, jak blisko przewidywania są rzeczywistych wartości.

In [52]:
model = Sequential()
model.add(Dense(units=32, activation='relu', input_shape=(150,)))
model.add(Dense(units=1))

model.compile(optimizer='rmsprop',
              loss='mae',
              metrics=['mse'])

model.fit(data, labels, epochs=30, batch_size=32, validation_split=0.2)

Epoch 1/30
25/25 ━━━━━━━━━━━━━━━━━━━━ 1s 9ms/step - loss: 21.7881 - mse: 665.5618 - val_loss: 16.0227 - val_mse: 385.4216
Epoch 2/30
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 14.8458 - mse: 333.7926 - val_loss: 12.4518 - val_mse: 228.8550
Epoch 3/30
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 12.8425 - mse: 228.2542 - val_loss: 11.6057 - val_mse: 189.7400
Epoch 4/30
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 12.0196 - mse: 195.6732 - val_loss: 11.5603 - val_mse: 185.8077
Epoch 5/30
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 11.3427 - mse: 179.1329 - val_loss: 11.5529 - val_mse: 185.5551
Epoch 6/30
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 11.5430 - mse: 183.1139 - val_loss: 11.5540 - val_mse: 185.6346
Epoch 7/30
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 11.8524 - mse: 192.4002 - val_loss: 11.5547 - val_mse: 185.9528
Epoch 8/30
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 11.7361 - mse: 189.9537 - val_loss: 11.5588 - val_mse: 186.0735
Epoch 9/30
25/25 ━━━━━━━

co oznaczają te wratośći z outputu metody model.fit()

9ms/step - loss: 21.7881 - mse: 665.5618 - val_loss: 16.0227 - val_mse: 385.4216

Gemini
Wartości, które widzisz w outputcie metody model.fit(), reprezentują różne metryki monitorowane podczas trenowania Twojego modelu. Oto ich znaczenie:

loss: To wartość funkcji straty (loss function) obliczona dla danych treningowych. Wskazuje, jak dobrze model radzi sobie z przewidywaniem wartości dla danych, na których jest trenowany. Niższa wartość loss oznacza lepsze dopasowanie modelu do danych treningowych. W Twoim przypadku, loss wynosi 21.7881.

mse: To Mean Squared Error (MSE) obliczony dla danych treningowych. MSE jest jedną z metryk używanych do oceny wydajności modeli regresji. Mierzy średni kwadrat błędu między przewidywaniami modelu a wartościami rzeczywistymi. Niższa wartość mse oznacza lepsze dopasowanie modelu. W Twoim przypadku, mse wynosi 665.5618.

val_loss: To wartość funkcji straty obliczona dla danych walidacyjnych. Dane walidacyjne to oddzielny zestaw danych, który nie jest używany do trenowania modelu. Służy do monitorowania wydajności modelu na niewidzianych wcześniej danych. Niższa wartość val_loss oznacza lepsze uogólnienie modelu, czyli lepsze radzenie sobie z nowymi danymi. W Twoim przypadku, val_loss wynosi 16.0227.

val_mse: To Mean Squared Error (MSE) obliczony dla danych walidacyjnych. Podobnie jak val_loss, niższa wartość val_mse oznacza lepsze uogólnienie modelu. W Twoim przypadku, val_mse wynosi 385.4216.

9ms/step: To czas potrzebny na przetworzenie jednego kroku (batch) danych treningowych. W Twoim przypadku, jeden krok zajmuje 9 milisekund.

Podsumowanie:

Te wartości dają Ci wgląd w to, jak dobrze Twój model uczy się i uogólnia na niewidzianych wcześniej danych. Monitorowanie tych metryk podczas treningu pozwala na ocenę postępów modelu i ewentualną modyfikację hiperparametrów w celu poprawy jego wydajności.

Zwróć uwagę na różnicę między metrykami obliczonymi dla danych treningowych (loss, mse) a metrykami obliczonymi dla danych walidacyjnych (val_loss, val_mse). Duża różnica między tymi wartościami może wskazywać na przeuczenie (overfitting) modelu, co oznacza, że model zbyt dobrze dopasował się do danych treningowych i słabo radzi sobie z nowymi danymi.

In [53]:
test_data = np.random.random((10, 150))

model.predict(test_data)

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 71ms/step


array([[23.995935],
       [21.668957],
       [23.033037],
       [23.64813 ],
       [21.178606],
       [20.31038 ],
       [25.928959],
       [25.432278],
       [20.9938  ],
       [20.5799  ]], dtype=float32)